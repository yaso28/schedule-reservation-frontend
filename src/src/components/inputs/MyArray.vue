<template>
  <MyFormGroup
    :label="label"
    :form="form"
    :field="field"
    :label-for="id"
  >
    <MyArrayNoWrapper
      :form="form"
      :field="field"
      :allows-empty="allowsEmpty"
      :distinct="distinct"
      :items-sorter="itemsSorter"
    >
      <template #default="{innerForm, innerField}">
        <slot
          :innerForm="innerForm"
          :innerField="innerField"
        />
      </template>
      <template #listItem="{item}">
        <slot
          name="listItem"
          :item="item"
        />
      </template>
    </MyArrayNoWrapper>
  </MyFormGroup>
</template>

<script>
export default {
  name: 'MyArray',
  components: {
    MyFormGroup: require('./MyFormGroup').default,
    MyArrayNoWrapper: require('./MyArrayNoWrapper').default
  },
  mixins: [require('./input-mixin').default],
  props: {
    allowsEmpty: {
      type: Boolean,
      default: false
    },
    distinct: {
      type: Boolean,
      default: true
    },
    itemsSorter: {
      type: [Function, Boolean],
      default: (item1, item2) => {
        if (item1 < item2) {
          return -1
        } else if (item1 > item2) {
          return 1
        } else {
          return 0
        }
      }
    }
  }
}
</script>
