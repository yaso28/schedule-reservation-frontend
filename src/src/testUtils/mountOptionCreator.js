import { createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import routes from '@/router/routes'
import Vuex from 'vuex'
import VueI18n from 'vue-i18n'
import i18nOptions from '@/i18n/options'
import VueMeta from 'vue-meta'
import CoreuiVue from '@coreui/vue'
import icons from '@/icons'
import services from '@/services'

class MountOptionCreator {
  create (storeModules = null) {
    const localVue = createLocalVue()

    // router
    localVue.use(VueRouter)
    const router = new VueRouter({
      mode: 'history',
      base: process.env.BASE_URL,
      routes
    })

    // i18n
    localVue.use(VueI18n)
    const i18n = new VueI18n(i18nOptions)

    // vue-meta
    localVue.use(VueMeta)

    // Core UI
    localVue.use(CoreuiVue)

    const option = {
      localVue,
      router,
      i18n,
      icons,
      provide: services
    }

    // store
    if (storeModules) {
      localVue.use(Vuex)
      option.store = new Vuex.Store({
        modules: storeModules
      })
    }

    return option
  }
}

export default new MountOptionCreator()
