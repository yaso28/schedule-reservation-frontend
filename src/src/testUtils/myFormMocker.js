import MyForm from '@/models/MyForm'
jest.mock('@/models/MyForm')

class MyFormMocker {
  createMockSingleField (field, initValue) {
    const input = {}
    input[field] = initValue
    return this.createMock(input)
  }

  createMock (input) {
    const mockForm = new MyForm(input)
    mockForm.values = input
    mockForm.errors = {}
    return mockForm
  }

  async setValue (wrapper, form, field, value) {
    await wrapper.setProps({
      form: Object.assign({}, form, {
        values: Object.assign({}, form.values, { [field]: value })
      })
    })
  }

  setValueNonProps (form, field, value) {
    form.values[field] = value
  }

  async setError (wrapper, form, field, errorMessage) {
    await wrapper.setProps({
      form: Object.assign({}, form, {
        errors: Object.assign({}, form.errors, { [field]: errorMessage })
      })
    })
  }
}

export default new MyFormMocker()
