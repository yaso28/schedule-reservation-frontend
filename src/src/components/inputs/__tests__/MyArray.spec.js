import { mount } from '@vue/test-utils'
import mountOptionCreator from '@/testUtils/mountOptionCreator'
import MyArray from '../MyArray'
import myFormMocker from '@/testUtils/myFormMocker'

jest.mock('@/models/MyForm')

const label = 'サンプル'
const field = 'sample_field'
const innerField = `${field}_inner`
const makeFormArgs = { [innerField]: null }

const item1 = 'item1'
const item2 = 'item2'
const item3 = 'item3'
const customItemsSorter = (item1, item2) => {
  if (item1 < item2) {
    return 1
  } else if (item1 > item2) {
    return -1
  } else {
    return 0
  }
}

const resetFormMock = (form, innerForm) => {
  form.setValue.mockReset()
  innerForm.reset.mockReset()
}
const addItem = async (wrapper, form, innerForm, item, expectedValueToSet = null) => {
  resetFormMock(form, innerForm)
  myFormMocker.setValueNonProps(innerForm, innerField, item)
  await wrapper.find('button.btn-array-add').trigger('click')
  if (expectedValueToSet) {
    await myFormMocker.setValue(wrapper, form, field, expectedValueToSet)
  }
}
const deleteItem = async (wrapper, form, innerForm, index, expectedValueToSet = null) => {
  resetFormMock(form, innerForm)
  await wrapper.findAll('button.btn-array-delete').at(index).trigger('click')
  if (expectedValueToSet) {
    await myFormMocker.setValue(wrapper, form, field, expectedValueToSet)
  }
}
const assertItemsList = wrapper => {
  expect(wrapper.find('ul.my-array-items').element).toMatchSnapshot()
}

describe('MyArray', () => {
  let form
  let innerForm
  let mountOption
  let apiService
  let wrapper
  beforeEach(() => {
    form = myFormMocker.createMockSingleField(field, [])
    innerForm = myFormMocker.createMock(makeFormArgs)

    mountOption = mountOptionCreator.create()
    mountOption.propsData = { label, form, field }
    mountOption.scopedSlots = {
      default: '<div>slot default. inner field = {{ props.innerField }}</div>'
    }

    apiService = mountOption.provide.apiService
    apiService.makeForm = jest.fn().mockReturnValueOnce(innerForm)
  })

  afterEach(() => {
    jest.resetAllMocks()
  })

  describe('通常', () => {
    beforeEach(() => {
      wrapper = mount(MyArray, mountOption)
    })

    test('内部フォーム作成', () => {
      expect(apiService.makeForm).toHaveBeenCalledWith(makeFormArgs)
    })

    test('表示', () => {
      expect(wrapper.element).toMatchSnapshot()
    })

    describe('バリデーションエラー', () => {
      beforeEach(async () => {
        await myFormMocker.setError(wrapper, form, field, 'エラーです')
      })

      test('表示', () => {
        expect(wrapper.element).toMatchSnapshot()
      })
    })

    describe('空を入力して追加ボタンクリック', () => {
      beforeEach(async () => {
        await addItem(wrapper, form, innerForm, '')
      })

      test('form.setValue()は呼び出さない', () => {
        expect(form.setValue).not.toHaveBeenCalled()
      })

      test('innerForm.reset()は呼び出さない', () => {
        expect(innerForm.reset).not.toHaveBeenCalled()
      })
    })

    describe('値を入力して追加ボタンクリック', () => {
      let expectedValueToSet
      beforeEach(async () => {
        expectedValueToSet = [item1]
        await addItem(wrapper, form, innerForm, item1, expectedValueToSet)
      })

      test('form.setValue()を呼出', () => {
        expect(form.setValue).toHaveBeenCalledWith(field, expectedValueToSet)
      })

      test('innerForm.reset()を呼出', () => {
        expect(innerForm.reset).toHaveBeenCalledWith()
      })

      describe('同じ値を入力して追加ボタンクリック', () => {
        beforeEach(async () => {
          await addItem(wrapper, form, innerForm, item1)
        })

        test('form.setValue()は呼び出さない', () => {
          expect(form.setValue).not.toHaveBeenCalled()
        })

        test('innerForm.reset()は呼び出さない', () => {
          expect(innerForm.reset).not.toHaveBeenCalled()
        })
      })

      describe('異なる値を入力して追加ボタンクリック', () => {
        beforeEach(async () => {
          expectedValueToSet = [item1, item3]
          await addItem(wrapper, form, innerForm, item3, expectedValueToSet)
        })

        test('form.setValue()を呼出', () => {
          expect(form.setValue).toHaveBeenCalledWith(field, expectedValueToSet)
        })

        test('innerForm.reset()を呼出', () => {
          expect(innerForm.reset).toHaveBeenCalledWith()
        })

        describe('中間の値を入力して追加ボタンクリック', () => {
          beforeEach(async () => {
            expectedValueToSet = [item1, item2, item3]
            await addItem(wrapper, form, innerForm, item2, expectedValueToSet)
          })

          test('form.setValue()を呼出', () => {
            expect(form.setValue).toHaveBeenCalledWith(field, expectedValueToSet)
          })

          test('innerForm.reset()を呼出', () => {
            expect(innerForm.reset).toHaveBeenCalledWith()
          })

          test('リスト表示', () => {
            assertItemsList(wrapper)
          })

          describe.each([
            [0, [item2, item3]],
            [1, [item1, item3]],
            [2, [item1, item2]]
          ])('削除ボタンクリック（index：%i）', (btnIndex, expectedValue) => {
            beforeEach(async () => {
              await deleteItem(wrapper, form, innerForm, btnIndex, expectedValue)
            })

            test('form.setValue()を呼出', () => {
              expect(form.setValue).toHaveBeenCalledWith(field, expectedValue)
            })

            test('innerForm.reset()は呼び出さない', () => {
              expect(innerForm.reset).not.toHaveBeenCalled()
            })
          })
        })
      })
    })
  })

  describe('初期値：あり', () => {
    beforeEach(() => {
      myFormMocker.setValueNonProps(form, field, [item1, item2])
      wrapper = mount(MyArray, mountOption)
    })

    test('リスト表示', () => {
      assertItemsList(wrapper)
    })

    describe('値を入力して追加ボタンクリック', () => {
      let expectedValueToSet
      beforeEach(async () => {
        expectedValueToSet = [item1, item2, item3]
        await addItem(wrapper, form, innerForm, item3, expectedValueToSet)
      })

      test('form.setValue()を呼出', () => {
        expect(form.setValue).toHaveBeenCalledWith(field, expectedValueToSet)
      })
    })
  })

  describe('slot', () => {
    describe('listItem', () => {
      beforeEach(() => {
        mountOption.scopedSlots.listItem = '<span class="slot-customized">{{ props.item }}</span>'
        wrapper = mount(MyArray, mountOption)
      })

      describe('値をセット', () => {
        beforeEach(async () => {
          await myFormMocker.setValue(wrapper, form, field, [item2])
        })

        test('リスト表示', () => {
          assertItemsList(wrapper)
        })
      })
    })
  })

  describe('オプション', () => {
    describe('allowsEmpty=true', () => {
      beforeEach(() => {
        mountOption.propsData.allowsEmpty = true
        wrapper = mount(MyArray, mountOption)
      })

      describe('空を入力して追加ボタンクリック', () => {
        let expectedValueToSet
        beforeEach(async () => {
          expectedValueToSet = ['']
          await addItem(wrapper, form, innerForm, '', expectedValueToSet)
        })

        test('form.setValue()を呼出', () => {
          expect(form.setValue).toHaveBeenCalledWith(field, expectedValueToSet)
        })

        test('innerForm.reset()を呼出', () => {
          expect(innerForm.reset).toHaveBeenCalledWith()
        })
      })
    })

    describe('distinct=false', () => {
      beforeEach(() => {
        mountOption.propsData.distinct = false
        wrapper = mount(MyArray, mountOption)
      })

      describe('同じ値を入力して追加ボタンクリック', () => {
        let expectedValueToSet
        beforeEach(async () => {
          expectedValueToSet = [item1, item1]
          await addItem(wrapper, form, innerForm, item1, [item1])
          await addItem(wrapper, form, innerForm, item1, expectedValueToSet)
        })

        test('form.setValue()を呼出', () => {
          expect(form.setValue).toHaveBeenCalledWith(field, expectedValueToSet)
        })

        test('innerForm.reset()を呼出', () => {
          expect(innerForm.reset).toHaveBeenCalledWith()
        })
      })
    })

    describe.each([
      ['custom', customItemsSorter, [item3, item1], [item3, item2, item1]],
      ['false', false, [item1, item3], [item1, item3, item2]]
    ])('itemsSorter=%s', (testCase, optionValue, expectedValueFirst, expectedValueSecond) => {
      beforeEach(() => {
        mountOption.propsData.itemsSorter = optionValue
        wrapper = mount(MyArray, mountOption)
      })

      describe('値を入力して追加ボタンクリック（2回）', () => {
        beforeEach(async () => {
          await addItem(wrapper, form, innerForm, item1, [item1])
          await addItem(wrapper, form, innerForm, item3, expectedValueFirst)
        })

        test('form.setValue()を呼出', () => {
          expect(form.setValue).toHaveBeenCalledWith(field, expectedValueFirst)
        })

        describe('値を入力して追加ボタンクリック（さらに1回）', () => {
          beforeEach(async () => {
            await addItem(wrapper, form, innerForm, item2, expectedValueSecond)
          })

          test('form.setValue()を呼出', () => {
            expect(form.setValue).toHaveBeenCalledWith(field, expectedValueSecond)
          })
        })
      })
    })
  })
})
