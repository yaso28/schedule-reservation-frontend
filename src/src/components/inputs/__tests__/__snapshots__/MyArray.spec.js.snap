// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyArray slot listItem 値をセット リスト表示 1`] = `
<ul
  class="list-group my-array-items"
  role="list-items"
>
  <li
    class="list-group-item d-flex align-items-center py-1"
  >
    <span
      class="flex-grow-1"
    >
      <span
        class="slot-customized"
      >
        item2
      </span>
    </span>
     
    <button
      class="btn btn-array-delete btn-outline-info btn-sm"
      type="button"
    >
      <svg
        class="c-icon"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polygon
          class="ci-primary"
          fill="var(--ci-primary-color, currentColor)"
          points="427.314 107.313 404.686 84.687 256 233.373 107.314 84.687 84.686 107.313 233.373 256 84.686 404.687 107.314 427.313 256 278.627 404.686 427.313 427.314 404.687 278.627 256 427.314 107.313"
        />
      </svg>
    </button>
  </li>
</ul>
`;

exports[`MyArray 初期値：あり リスト表示 1`] = `
<ul
  class="list-group my-array-items"
  role="list-items"
>
  <li
    class="list-group-item d-flex align-items-center py-1"
  >
    <span
      class="flex-grow-1"
    >
      
          item1
        
    </span>
     
    <button
      class="btn btn-array-delete btn-outline-info btn-sm"
      type="button"
    >
      <svg
        class="c-icon"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polygon
          class="ci-primary"
          fill="var(--ci-primary-color, currentColor)"
          points="427.314 107.313 404.686 84.687 256 233.373 107.314 84.687 84.686 107.313 233.373 256 84.686 404.687 107.314 427.313 256 278.627 404.686 427.313 427.314 404.687 278.627 256 427.314 107.313"
        />
      </svg>
    </button>
  </li>
  <li
    class="list-group-item d-flex align-items-center py-1"
  >
    <span
      class="flex-grow-1"
    >
      
          item2
        
    </span>
     
    <button
      class="btn btn-array-delete btn-outline-info btn-sm"
      type="button"
    >
      <svg
        class="c-icon"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polygon
          class="ci-primary"
          fill="var(--ci-primary-color, currentColor)"
          points="427.314 107.313 404.686 84.687 256 233.373 107.314 84.687 84.686 107.313 233.373 256 84.686 404.687 107.314 427.313 256 278.627 404.686 427.313 427.314 404.687 278.627 256 427.314 107.313"
        />
      </svg>
    </button>
  </li>
</ul>
`;

exports[`MyArray 通常 バリデーションエラー 表示 1`] = `
<div
  class="form-row form-group"
>
  <label
    class="col-form-label col-sm-3"
    for="sample_field"
  >
    
    サンプル
  
  </label>
   
  <div
    class="col-sm-9"
  >
    <div
      class="my-input-wrapper is-invalid"
    >
      <div>
        <div
          class="d-flex"
        >
          <div
            class="flex-grow-1"
          >
            <div>
              slot default. inner field = sample_field_inner
            </div>
          </div>
           
          <button
            class="btn btn-array-add btn-outline-info"
            type="button"
          >
            
      追加
    
          </button>
        </div>
         
        <ul
          class="list-group my-array-items"
          role="list-items"
        />
      </div>
    </div>
     
    <div
      class="invalid-feedback"
    >
      
      エラーです
    
    </div>
  </div>
</div>
`;

exports[`MyArray 通常 値を入力して追加ボタンクリック 異なる値を入力して追加ボタンクリック 中間の値を入力して追加ボタンクリック リスト表示 1`] = `
<ul
  class="list-group my-array-items"
  role="list-items"
>
  <li
    class="list-group-item d-flex align-items-center py-1"
  >
    <span
      class="flex-grow-1"
    >
      
          item1
        
    </span>
     
    <button
      class="btn btn-array-delete btn-outline-info btn-sm"
      type="button"
    >
      <svg
        class="c-icon"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polygon
          class="ci-primary"
          fill="var(--ci-primary-color, currentColor)"
          points="427.314 107.313 404.686 84.687 256 233.373 107.314 84.687 84.686 107.313 233.373 256 84.686 404.687 107.314 427.313 256 278.627 404.686 427.313 427.314 404.687 278.627 256 427.314 107.313"
        />
      </svg>
    </button>
  </li>
  <li
    class="list-group-item d-flex align-items-center py-1"
  >
    <span
      class="flex-grow-1"
    >
      
          item2
        
    </span>
     
    <button
      class="btn btn-array-delete btn-outline-info btn-sm"
      type="button"
    >
      <svg
        class="c-icon"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polygon
          class="ci-primary"
          fill="var(--ci-primary-color, currentColor)"
          points="427.314 107.313 404.686 84.687 256 233.373 107.314 84.687 84.686 107.313 233.373 256 84.686 404.687 107.314 427.313 256 278.627 404.686 427.313 427.314 404.687 278.627 256 427.314 107.313"
        />
      </svg>
    </button>
  </li>
  <li
    class="list-group-item d-flex align-items-center py-1"
  >
    <span
      class="flex-grow-1"
    >
      
          item3
        
    </span>
     
    <button
      class="btn btn-array-delete btn-outline-info btn-sm"
      type="button"
    >
      <svg
        class="c-icon"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polygon
          class="ci-primary"
          fill="var(--ci-primary-color, currentColor)"
          points="427.314 107.313 404.686 84.687 256 233.373 107.314 84.687 84.686 107.313 233.373 256 84.686 404.687 107.314 427.313 256 278.627 404.686 427.313 427.314 404.687 278.627 256 427.314 107.313"
        />
      </svg>
    </button>
  </li>
</ul>
`;

exports[`MyArray 通常 表示 1`] = `
<div
  class="form-row form-group"
>
  <label
    class="col-form-label col-sm-3"
    for="sample_field"
  >
    
    サンプル
  
  </label>
   
  <div
    class="col-sm-9"
  >
    <div
      class="my-input-wrapper"
    >
      <div>
        <div
          class="d-flex"
        >
          <div
            class="flex-grow-1"
          >
            <div>
              slot default. inner field = sample_field_inner
            </div>
          </div>
           
          <button
            class="btn btn-array-add btn-outline-info"
            type="button"
          >
            
      追加
    
          </button>
        </div>
         
        <ul
          class="list-group my-array-items"
          role="list-items"
        />
      </div>
    </div>
     
    <!---->
  </div>
</div>
`;
